pennylane/tape/qnode.py

```diff
@@ -815,13 +847,32 @@ def circuit(weights):
     """
 
     def _metric_tensor_fn(*args, **kwargs):
+        try:
+            jac = qml.math.stack(_get_classical_jacobian(qnode)(*args, **kwargs))
+            jac = qml.math.reshape(jac, [qnode.qtape.num_params, -1])
+            perm = np.argsort(np.argsort(np.nonzero(qml.math.toarray(jac))[1]))
+        except:
+            perm = None
+
         qnode.construct(args, kwargs)
+
         metric_tensor_tapes, processing_fn = qml.tape.transforms.metric_tensor(
             qnode.qtape, diag_approx=diag_approx
         )
 
+        if only_construct:
+            return metric_tensor_tapes
+
         res = [t.execute(device=qnode.device) for t in metric_tensor_tapes]
-        return processing_fn(res)
+        metric_tensor = processing_fn(res)
+
+        if perm is None or np.all(perm == np.arange(qnode.qtape.num_params)):

```

Comment:
Note the `np.all` method will fail in `jit` contexts.

I would actually recommend just removing the second part of the check. Gather is usually smart enough to do nothing if it doesn't have to.
```suggestion
        if perm is None:

```